app/src/main/java/co/kaush/msusf/usf/UsfVmImpl.kt

```diff
@@ -1,74 +1,97 @@
+    override fun warning(message: String) = Timber.w(message)
+    override fun error(error: Throwable, message: String) = Timber.e(error, message)
+  }
 ) : UsfVm<E, R, VS, VE> {
 
-    override val eventSink: PublishSubject<E> = PublishSubject.create()
-    override val disposables: CompositeDisposable = CompositeDisposable()
+  // remember that a single E(vent) can yield multiple R(esult)s
+  //  so if you wish to show an initial loading state before the final result
+  //  emit multiple R(esult)s for the single E(vent)
+  abstract fun eventToResultFlow(): (E) -> Flow<R>
 
-    private val viewStateSink: Observable<VS>
-    private val viewEffectSink: Observable<VE>
+  // curiously, we don't return a Flow<VS> here
+  //  as every (R)esult will only ever be transformed into a single (V)iew(S)tate
+  //  if you want multiple (V)iew(S)tates, then emit multiple (R)esults
+  //  and transform each (R)esult accordingly to the respective (V)iew(S)tate
+  abstract fun resultToViewState(): (currentViewState: VS, result: R) -> VS
 
-    init {
-        logger.debug("------ init ${Thread.currentThread().name}")
+  // a single (R)esult can result in multiple (V)iew(E)ffects
+  //  for e.g. emit one VE for navigation and another for making an analytics call
+  //  hence a return type of Flow<VE>
+  abstract fun resultToViewEffectFlow(): (R) -> Flow<VE>
 
-        eventSink
-            .compose(eventToResultTransformer)
-            .doOnNext { logger.debug("----- result ${Thread.currentThread().name} $it") }
+  abstract val initialViewState: VS
+  private val eventSink = MutableSharedFlow<E>()
+  private lateinit var viewStateSink: Flow<VS>
+  private lateinit var viewEffectSink: Flow<VE>
 
-            // share the result stream otherwise it will get subscribed to multiple times
-            // in the following also block
-            .share()
 
-            .also { result ->
-                 logger.debug("------ also ${Thread.currentThread().name}")
-                viewStateSink = result
-                    .compose(resultToViewStateTransformer)
+  init {
+    logger.debug("------ init ${Thread.currentThread().name}")
 
+    viewModelScope.launch {
+      val resultFlow: Flow<R> = eventSink.flatMapLatest(transform = eventToResultFlow())
+          .onEach { logger.debug("----- result ${Thread.currentThread().name} $it") }
 
-                    // if the viewState is identical
-                    // there's little reason to re-emit the same view state
-                    .distinctUntilChanged()
+      // share the result stream otherwise it will get subscribed to multiple times
+      // in the following also block
+      //  TODO: i'm not yet sure if we need this given we're using a mutable shared flow
 
-                    .doOnNext { logger.debug("----- vs $it") }
+      withContext(processingDispatcher) {
+        viewStateSink = resultFlow
+            .scan(initialViewState, resultToViewState())
```

Comment:
i actually think this might be a big improvement in the existing implementation of this pattern, as we're now enforcing the `scan` operator.

this means, we don't expect the implementation to remember to use the logic this way. it enforces it.